src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs

```diff
@@ -1,207 +1,375 @@
+                _sessionMessageProcessor = _messagingProvider.CreateSessionMessageProcessor(_entityPath, _serviceBusAccount.ConnectionString);
+            }
+            else
+            {
+                _messageProcessor = _messagingProvider.CreateMessageProcessor(entityPath, _serviceBusAccount.ConnectionString);
+            }
+            _serviceBusOptions = config;
+        }
+
+        internal MessageReceiver Receiver => _receiver.Value;
+
+        internal IMessageSession MessageSession => _messageSession;
+
+        public Task StartAsync(CancellationToken cancellationToken)
+        {
+            ThrowIfDisposed();
+
+            if (_started)
+            {
+                throw new InvalidOperationException("The listener has already been started.");
+            }
+
+            if (_receiver == null)
+            {
+                _receiver = CreateMessageReceiver();
+            }
+
+            if (_isSessionsEnabled)
+            {
+                _clientEntity = _messagingProvider.CreateClientEntity(_entityPath, _serviceBusAccount.ConnectionString);
+
+                if (_singleDispatch)
+                {
+                    if (_clientEntity is QueueClient queueClient)
+                    {
+                        queueClient.RegisterSessionHandler(ProcessSessionMessageAsync, _serviceBusOptions.SessionHandlerOptions);
+                    }
+                    else
+                    {
+                        SubscriptionClient subscriptionClient = _clientEntity as SubscriptionClient;
+                        subscriptionClient.RegisterSessionHandler(ProcessSessionMessageAsync, _serviceBusOptions.SessionHandlerOptions);
+                    }
+                }
+                else
+                {
+                    StartMultipleMessageListening(_cancellationTokenSource.Token);
+                }
+            }
+            else
+            {
+                if (_singleDispatch)
+                {
+                    Receiver.RegisterMessageHandler(ProcessMessageAsync, _serviceBusOptions.MessageHandlerOptions);
+                }
+                else
+                {
+                    StartMultipleMessageListening(_cancellationTokenSource.Token);
+                }
+            }
+            _started = true;
+
+            return Task.CompletedTask;
+        }
+
+        public async Task StopAsync(CancellationToken cancellationToken)
+        {
+            ThrowIfDisposed();
+
+            if (!_started)
+            {
+                throw new InvalidOperationException("The listener has not yet been started or has already been stopped.");
+            }
+
+            // cancel our token source to signal any in progress
+            // ProcessMessageAsync invocations to cancel
+            _cancellationTokenSource.Cancel();
+
+            if (_receiver != null && _receiver.IsValueCreated)
+            {
+                await Receiver.CloseAsync();
+                _receiver = null;

```

Comment:
@gzuber made the code this way in his [recent commit](https://github.com/Azure/azure-functions-servicebus-extension/commit/abc92e91bdc02b36a2e6ceae8f0b03ee65e747dd), I think to ensure the Lazy was initialized?

Reply:
Oh it looks like @alrod added the `CreateMessageReceiver` call to `StartAsync` so that should work (making the assumption that nothing can be called between a Stop and subsequent Start)

Reply:
I moved back `_receiver = CreateMessageReceiver();` in `StopAsync`